File: chapter 2/2.5.py
# 2.5 Sum Lists: You have two numbers represented by a linked list, where each node contains a single
# digit. The digits are stored in reverse order, such that the Vs digit is at the head of the list. Write a
# function that adds the two numbers and returns the sum as a linked list.
# EXAMPLE
# Input: (7- > 1 -> 6) + (5 -> 9 -> 2).That is,617 + 295.
# Output: 2 -> 1 -> 9. That is, 912.
# FOLLOW UP
# Suppose the digits are stored in forward order. Repeat the above problem.
# EXAMPLE
# Input: (6 -> 1 -> 7) + (2 -> 9 -> 5).That is, 617 + 295,
# Output:9 -> 1 -> 2,Thatis,912.
# Hints: #7, #30, #71 #95, #109 

class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
    
class SLL:
    def __init__ (self):
        self.head = None

    def printAll(self):
        current = self.head
        while current != None:
            print(current.val)
            current = current.next
    
    def addNode(self, val):
        if self.head == None:
            self.head = Node(val)
        else:
            newNode = Node(val)
            current = self.head
            while current.next != None:
                current = current.next
            current.next = newNode
    
    def removeFromMiddle(self, val):
        if self.head == None or self.head.next == None:
            return False
        current = self.head
        while current.next != None:
            if current.next.val == val and current.next.next != None:
                current.next = current.next.next
                isRemoved = True
                return 
            current = current.next
        print("the value does not exist")

def sumLists (SLL1, SLL2):
    if SLL1.head == None and SLL2.head == None:
        return None
    elif SLL1.head == None:
        return SLL2
    elif SLL2.head == None:
        return SLL1
    else:
        current = SLL1.head
        power = 0
        number1 = 0
        while (current != None):
            number1 += (current.val * (10**power))
            power += 1
            current = current.next
        
        current = SLL2.head
        power = 0
        number2 = 0
        while (current != None):
            number2 += (current.val * (10**power))
            power += 1
            current = current.next
        sum = number1 + number2
        strSum = str(sum)
        print(strSum)
        newList = SLL()
        for i in range(len(strSum)-1, -1, -1):
            newList.addNode(strSum[i])
        return newList.printAll()

SLL1 = SLL()
SLL1.addNode(7) 
SLL1.addNode(1) 
SLL1.addNode(6) 

SLL2 = SLL()
SLL2.addNode(5)
SLL2.addNode(9)
SLL2.addNode(2)

sumLists(SLL1,SLL2)

File: chapter 2/js/2.5.js
// Sum Lists: You have two numbers represented by a linked list, where each node contains a single
// digit. The digits are stored in reverse order, such that the Vs digit is at the head of the list. Write a
// function that adds the two numbers and returns the sum as a linked list.
// EXAMPLE
// Input: (7- > 1 -> 6) + (5 -> 9 -> 2).That is,617 + 295.
// Output: 2 -> 1 -> 9. That is, 912.
// FOLLOW UP
// Suppose the digits are stored in forward order. Repeat the above problem.
// EXAMPLE
// Input: (6 -> 1 -> 7) + (2 -> 9 -> 5).That is, 617 + 295, 

class Node{
    constructor(value){
        this.value = value;
        this.next = null 
    }
}

class SLL{
    constructor(){
        this.head = null;
    }

    append(val){
        let newNode = new Node(val);
        if (this.head == null){
            this.head = newNode
            return
        }
        let runner = this.head;
        while(runner.next != null){
            runner = runner.next;
        }   
        runner.next = newNode;
        return
    }
    printAll(){
        let runner = this.head
        while(runner != null){
            console.log(runner.value)
            runner = runner.next
        }
    }
}
//reverse order
// function sumLists(list1,list2){
//     if(list1.head == null && list2.head == null){
//         return null;
//     } else if (list1.head == null || list2.head == null){
//         return list1.head == null? list2 : list1
//     }
//     let runner = list1.head;
//     let str1="";
//     let str2="";
//     while(runner != null){
//         str1 = runner.value + str1;
//         runner = runner.next
//     }
//     runner = list2.head;
//     while(runner != null){
//         str2 = runner.value + str2;
//         runner = runner.next
//     }
//     console.log(str1)
//     console.log(str2)
//     let sum = (+str1) + (+str2)
//     let sumStr = sum.toString()
//     let sumList = new SLL();
//     for(let i=sumStr.length-1; i>=0; i--){
//         sumList.append(sumStr[i])
//     }
//     console.log(sumList)
//     return sumList
// }

//forward order
function sumLists(list1,list2){
    if(list1.head == null && list2.head == null){
        return null;
    } else if (list1.head == null || list2.head == null){
        return list1.head == null? list2 : list1
    }
    let runner = list1.head;
    let str1="";
    let str2="";
    while(runner != null){
        str1 = str1 + runner.value;
        runner = runner.next
    }
    runner = list2.head;
    while(runner != null){
        str2 = str2 + runner.value;
        runner = runner.next
    }
    console.log(str1)
    console.log(str2)
    let sum = (+str1) + (+str2)
    let sumStr = sum.toString()
    let sumList = new SLL();
    for(let i=0; i<sumStr.length; i++){
        sumList.append(sumStr[i])
    }
    console.log(sumList)
    return sumList
}



let sll1 = new SLL()
sll1.append(6)
sll1.append(1)
sll1.append(7)
let sll2 = new SLL()
sll2.append(2)
sll2.append(9)
sll2.append(5)

let list = sumLists(sll1, sll2)
console.log(list)
list.printAll()
